pkg/networkservice/common/monitor/server.go

```diff
@@ -77,43 +85,80 @@ func (m *monitorServer) MonitorConnections(selector *networkservice.MonitorScope
 }
 
 func (m *monitorServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	// Pass metrics monitor, so it could be used later in chain.
+	ctx = WithServer(ctx, m)
+
+	connectionClone := proto.Clone(request.GetConnection())
 	conn, err := next.Server(ctx).Request(ctx, request)
 	if err == nil {
 		m.executor.AsyncExec(func() {
 			m.connections[conn.GetId()] = conn
-			event := &networkservice.ConnectionEvent{
-				Type:        networkservice.ConnectionEventType_UPDATE,
-				Connections: map[string]*networkservice.Connection{conn.GetId(): conn},
+
+			// Send update only if connection is updated.
+			if !proto.Equal(connectionClone, conn) {
+				event := &networkservice.ConnectionEvent{
+					Type:        networkservice.ConnectionEventType_UPDATE,
+					Connections: map[string]*networkservice.Connection{conn.GetId(): conn},
+				}
+				if err = m.send(ctx, event); err != nil {
+					logrus.Errorf("Error during sending event: %v", err)
+				}
 			}
-			m.send(ctx, event)
 		})
 	}
 	return conn, err
 }
 
 func (m *monitorServer) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
+	// Pass metrics monitor, so it could be used later in chain.
+	ctx = WithServer(ctx, m)
+
 	m.executor.AsyncExec(func() {
 		delete(m.connections, conn.GetId())
 		event := &networkservice.ConnectionEvent{
 			Type:        networkservice.ConnectionEventType_DELETE,
 			Connections: map[string]*networkservice.Connection{conn.GetId(): conn},
 		}
-		m.send(ctx, event)
+		if err := m.send(ctx, event); err != nil {
+			logrus.Errorf("Error during sending event: %v", err)
+		}
 	})
 	return next.Server(ctx).Close(ctx, conn)
 }
 
-func (m *monitorServer) send(ctx context.Context, event *networkservice.ConnectionEvent) {
-	newMonitors := make([]networkservice.MonitorConnection_MonitorConnectionsServer, len(m.monitors))
-	for _, srv := range m.monitors {
-		select {
-		case <-srv.Context().Done():
-		default:
-			if err := srv.Send(event); err != nil {
-				trace.Log(ctx).Errorf("Error sending event: %+v: %+v", event, err)
+func (m *monitorServer) Send(event *networkservice.ConnectionEvent) error {
+	m.executor.AsyncExec(func() {
```

Comment:
This method already called from `AsyncExec` why we need to do it again?

Reply:
We could not be sure.